### PR TITLE
feat: Refactor rest-api

### PR DESCRIPTION
**Pull request by Abdul-RehmanMalik**
**_"Refactored Code"_**
**Files Updated:** 

- controllers/authControllers.ts 
- routes/authRoutes.ts
- routes/index.ts
- models/User.ts
- src/app.ts

**Files Added:**

- controllers/userControllers.ts 
- routes/userRoutes.ts
- util/generateAccessToken.ts
- util/Validation.ts

### CODE REVIEW

File: src/routes/userRoutes.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import { getUserValidation } from "../util/validation";
+import { UserController } from "../controllers/userController";
+
+const userRouter = express.Router();
+const userController = new UserController();
+// getuser route
+userRouter.get("/getuser", async (req, res) => {
+  const { error, value: params } = getUserValidation(req.params);
+  const { username } = params;
+  if (error) return res.status(400).send(error.details[0].message);
```
Comment (sabhas) [inline src/routes/userRoutes.ts:11]: move this line above const {username} = params

Comment (sabhas) [inline src/routes/userRoutes.ts:11]: move this line above const {username} = params

File: src/routes/authRoutes.ts [hunk 2]
```diff
@@ -1,10 +1,30 @@
...
-// SignUp route
-authRouter.post("/signup", createUser);
+const authRouter = express.Router();
+const authController = new AuthController();
+// Signup route
+authRouter.post("/signup", async (req, res) => {
+  const { error, value: body } = signUpValidation(req.body);
+  if (error) return res.status(400).send(error.details[0].message);
+  try {
+    const response = await authController.createUser(body);
```
Comment (sabhas) [inline src/routes/authRoutes.ts:12]: Although we are creating users on sign-up. it is better to have consistent names for actions. So, i would suggest to change controllers method name to signup from createuser

File: src/controllers/authController.ts [hunk 3]
```diff
@@ -1,58 +1,103 @@
...
+
+    if (!user) {
+      throw {
+        code: 401, //401 Unauthorized is the status code to return when the client provides no credentials or invalid credentials.
+        message: "Invalid Email or Password",
+      };
+    }
+
+    // Check password validity
+    const isPasswordValid = await bcrypt.compare(password, user.password);
```
Comment (sabhas) [inline src/controllers/authController.ts:77]: You can perform password comparison and password hashing in the user model. You need to create static methods in the model for these actions.

**Review by sabhas — CHANGES_REQUESTED** [review]
you need to improve file nomenclature. there's no need to specify controllers or routes in file names. Let's say a folder named controller contains some files and it will be obvious that these are controller files. So you can just name it user.ts instead of userController.ts. The same applies for every other folder.

File: src/models/User.ts [hunk 4]
```diff
@@ -1,6 +1,27 @@
...
+   */
+  password: string;
+  /**
+   * address for user
+   * @example "H#123 Block 2 Sector J, Abc Town, NY"
+   */
+  address: string;
+}
+interface UserDocument extends UserPayload, Document {
   name: string;
```
Comment (sabhas) [inline src/models/User.ts:25]: you are extending UserPayload and all these properties exist in that interface. So, there's no need to specify these again on this UserDocument

**Review by sabhas — CHANGES_REQUESTED** [review]

File: src/routes/userRoutes.ts (hunk 1 above)
Comment (Abdul-RehmanMalik) [reply src/routes/userRoutes.ts]: done

File: src/routes/authRoutes.ts (hunk 2 above)
Comment (Abdul-RehmanMalik) [reply src/routes/authRoutes.ts]: doone

File: src/controllers/authController.ts (hunk 3 above)
Comment (Abdul-RehmanMalik) [reply src/controllers/authController.ts]: done

File: src/models/User.ts (hunk 4 above)
Comment (Abdul-RehmanMalik) [reply src/models/User.ts]: done

File: package.json [hunk 5]
```diff
@@ -5,9 +5,10 @@
   "main": "index.js",
   "scripts": {
     "test": "echo \"Error: no test specified\" && exit 1",
-    "start": "tsc && nodemon dist/app.js",
     "prettier": "prettier --write \"src/**/*.ts\"",
-    "lint": "eslint --ext .ts ."
+    "lint": "eslint --ext .ts .",
+    "start": "nodemon dist/app.js",
```
Comment (sabhas) [inline package.json:10]: update this script

File: package.json [hunk 6]
```diff
@@ -16,20 +17,29 @@
...
+    "@types/cors": "^2.8.13",
     "@types/express": "^4.17.17",
     "@types/jsonwebtoken": "^9.0.2",
-    "@types/node": "^20.1.0",
+    "@types/node": "^20.1.3",
+    "@types/swagger-ui-express": "^4.1.3",
     "@typescript-eslint/eslint-plugin": "^5.59.5",
     "@typescript-eslint/parser": "^5.59.5",
     "eslint": "^8.40.0",
+    "express": "^4.18.2",
```
Comment (sabhas) [inline package.json:39]: express is not a dev dependency at all.

File: src/routes/users.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
+import express from "express";
+import { getUserValidation } from "../util/validation";
+import { UserController } from "../controllers/users";
+
+const userRouter = express.Router();
+const userController = new UserController();
+// getuser route
+userRouter.get("/getuser", async (req, res) => {
```
Comment (sabhas) [inline src/routes/users.ts:8]: there's no need to explicitly add `getuser` in api endpoint. HTTP GET `user/` already implies that you want to get a user.

**Review by sabhas — CHANGES_REQUESTED** [review]

File: package.json (hunk 5 above)
Comment (Abdul-RehmanMalik) [reply package.json:10]: done

File: package.json (hunk 6 above)
Comment (Abdul-RehmanMalik) [reply package.json:39]: done

File: src/routes/users.ts (hunk 7 above)
Comment (Abdul-RehmanMalik) [reply src/routes/users.ts:8]: done

**Pull request merged**